### DISCUSSION TITLE
Using PIXI JS with THREE

### DISCUSSION THREAD

**Discussion by nickyvanurk** [pixijs/pixijs#11138] [Q&A]
Hi! I am trying to use pixi to render the HUD in my game, but I use three.js for 3D graphics. How would I go about combining them? I found this (https://github.com/pixijs/pixijs/discussions/7767) old discussion but the provided answer no longer works. I've gone through the pixi docs (v8) and experimented with different settings/function calls/classes but I can't figure it out by myself. Please help :sob:

**Comment by nickyvanurk** [comment]
I figured it out: https://jsfiddle.net/kotp27x4/

**Comment by GoodBoyDigital** [comment]
hey @nickyvanurk!

This line:

` pRenderer.runners.contextChange.emit(pRenderer.gl);`

will force Pixi to recreate the GPU reousrces each frame (without cleaning up the old ones). 

ideally we need to build a decent `resetGLState` function into v8 that resets all the caches so that it can render three stuff wthout weirdness. I'll see if I can get to that sooner rather than later!

**Comment by GoodBoyDigital** [comment]
here we go! https://github.com/pixijs/pixijs/pull/11142

**Comment by nickyvanurk** [comment]
Thank you so much! Even added a guide uh 🙏Will help a ton of people I'm sure. The current info online is severely outdated.

**Comment by nickyvanurk** [comment]
@GoodBoyDigital Is there a problem with the change you made? For me pixijs is unusable. It progressively gets more slow until it takes more than 2 seconds to render a frame. One thing that takes a lot of time is pixijs trying to get the width of the container for some reason (which I am not doing anywhere in my code).
![image](https://github.com/user-attachments/assets/f16982b6-b606-4470-b287-cbbe262fc396)
That refers to this code here: https://github.com/pixijs/pixijs/blob/9d177e2d2d3c33df2b0ae7afef180c5028e4d3af/src/scene/container/Container.ts#L1033

And the render function itself takes even longer (2.5s or so with just 11 elements in the scene).
https://github.com/pixijs/pixijs/blob/9d177e2d2d3c33df2b0ae7afef180c5028e4d3af/src/rendering/renderers/shared/system/AbstractRenderer.ts#L261

I checked any problems in my own code but I have consistently just 11 elements in my pixijs scene so I really don't know what the problem could be, I suspect pixijs has a bug.

Do you mind checking if something is wrong with pixi.js? As mentioned pixijs is unusable for me right now (you can try it out at http://everwilds.io/ press v to toggle nameplates. With nameplates turned on the game gets progressively slower).

This is how I create my pixi scene and render it:

![image](https://github.com/user-attachments/assets/06aeff2f-b119-4654-b84a-51c7ae5cd2e6)

![image](https://github.com/user-attachments/assets/1d849f84-3fc7-45fa-95b4-802266fa7dfa)

**Reply by GoodBoyDigital** [reply]
This kind of thing usually suggests a scene is forever growing at a very high rate. Have you tried inspecting the scene with Pixi debug tool?

**Comment by nickyvanurk** [comment]
@GoodBoyDigital 

The scene is not growing it seems. You can try the devtools yourself at http://everwilds.io

What seems to be odd is the size of the pixi.js scene seems to get changed whenever I change my three.js orbit controls camera angle. The size is also really huge which is strange. Could that be the source of the problem? Nothing else seems to indicate something growing.

![image](https://github.com/user-attachments/assets/e7ddb0a2-9144-4182-81fc-ef0cf4f9bdc6)

The only time I set the pixi.js renderer size is when I init it as seen in my previous screenshot

**Comment by nickyvanurk** [comment]
It seems the size of the scene gets changed depending on the location of the elements in the pixi scene. As my monsters are off-screen and still have a HUD the scene can get really big. When I move away to make sure the pixi scene is small that improved the render time by a lot. But that did not fix the slowdown issue. Doing a performance monitor really shows it's pixi that's the cause somehow. I even tried to remove the text HUD elements so it just shows the health bars.

![image](https://github.com/user-attachments/assets/643fac70-9831-46b5-98c1-0251beb709a8)

Note that this is all happening on a beefy laptop with a AMD Ryzen 9 7945HX  and RTX 4070
